src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+          a (Z)Catalog instance
+        """
+
+        self.id = id
+        self.ignore_ex = ignore_ex        # currently unimplimented
+        self.call_methods = call_methods
+
+        self.operators = ('or', 'and')
+        self.useOperator = 'or'
+
+        # set components
+        self._components = ComponentMapping()
+        if extra:
+            for cdata in extra:
+                c_id = cdata['id']
+                c_meta_type = cdata['meta_type']
+                c_attributes = cdata['attributes']
+                self._components[c_id] = Component(c_id, c_meta_type,
+                                                   c_attributes)
+        self.clear()
+
+    def _apply_index(self, request, resultset=None):
+        """ Apply the index to query parameters given in the request arg. """
+        record = parseIndexRequest(request, self.id, self.query_options)
+
+        if record.keys is None:
+            return None
+
+        return UnIndex._apply_index(self, request, resultset=resultset)
+
+    def index_object(self, documentId, obj, threshold=None):
+        """ wrapper to handle indexing of multiple attributes """
+
+        res = self._index_object(documentId, obj, threshold)
+        return res
+
+    def _index_object(self, documentId, obj, threshold=None):
+        """ index an object 'obj' with integer id 'i'
+
+        Ideally, we've been passed a sequence of some sort that we
+        can iterate over. If however, we haven't, we should do something
+        useful with the results. In the case of a string, this means
+        indexing the entire string as a keyword."""
+
+        # First we need to see if there's anything interesting to look at
+        # self.id is the name of the index, which is also the name of the
+        # attribute we're interested in.  If the attribute is callable,
+        # we'll do so.
+
+        # get permuted keywords
+        newKeywords = self._get_object_keywords(obj)
+
+        oldKeywords = self._unindex.get(documentId, None)
+
+        if oldKeywords is None:
+            # we've got a new document, let's not futz around.
+            try:
+                for kw in newKeywords:
+                    self.insertForwardIndexEntry(kw, documentId)
+                if newKeywords:
+                    self._unindex[documentId] = list(newKeywords)
+            except TypeError:
+                return 0
+        else:
+            # we have an existing entry for this document, and we need
+            # to figure out if any of the keywords have actually changed
+            if type(oldKeywords) is not OOSet:
+                oldKeywords = OOSet(oldKeywords)
+            newKeywords = OOSet(newKeywords)
+            fdiff = difference(oldKeywords, newKeywords)
+            rdiff = difference(newKeywords, oldKeywords)
+            if fdiff or rdiff:
+                # if we've got forward or reverse changes
+                self._unindex[documentId] = list(newKeywords)
```

Comment:
KeywordIndex._index_object has additional logic for deletes here, maybe those should also go here:
```
# if we've got forward or reverse changes
if newKeywords:
    self._unindex[documentId] = list(newKeywords)
else:
    del self._unindex[documentId]
if fdiff:
...
```